File: frontend/src/lib/api/records.ts
import { apiFetch } from "./client";
import type { DnsRecord, ListParams, PaginatedResponse } from "@/types";

function buildQuery(params: ListParams = {}): string {
  const searchParams = new URLSearchParams();
  if (params.page) searchParams.set("page", String(params.page));
  if (params.page_size) searchParams.set("page_size", String(params.page_size));
  if (params.search) searchParams.set("search", params.search);
  if (params.type) searchParams.set("type", params.type);
  if (params.sort_by) searchParams.set("sort_by", params.sort_by);
  if (params.sort_order) searchParams.set("sort_order", params.sort_order);
  const qs = searchParams.toString();
  return qs ? `?${qs}` : "";
}

export async function listRecords(
  zoneId: number,
  params: ListParams = {},
): Promise<PaginatedResponse<DnsRecord>> {
  return apiFetch<PaginatedResponse<DnsRecord>>(
    `/hosted-zones/${zoneId}/records${buildQuery(params)}`,
  );
}

export async function createRecord(
  zoneId: number,
  data: {
    name: string;
    type: string;
    ttl: number;
    value: string;
    routing_policy?: string;
    comment?: string;
  },
): Promise<DnsRecord> {
  return apiFetch<DnsRecord>(`/hosted-zones/${zoneId}/records`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateRecord(
  id: number,
  data: {
    name?: string;
    type?: string;
    ttl?: number;
    value?: string;
    routing_policy?: string;
    comment?: string;
  },
): Promise<DnsRecord> {
  return apiFetch<DnsRecord>(`/records/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export async function deleteRecord(id: number): Promise<void> {
  return apiFetch<void>(`/records/${id}`, { method: "DELETE" });
}

export async function bulkDeleteRecords(
  zoneId: number,
  recordIds: number[],
): Promise<{ deleted: number }> {
  return apiFetch<{ deleted: number }>(
    `/hosted-zones/${zoneId}/records/bulk-delete`,
    {
      method: "POST",
      body: JSON.stringify({ record_ids: recordIds }),
    },
  );
}

File: backend/app/routers/records.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.routers.auth import get_current_user
from app.schemas import DnsRecordCreate, DnsRecordResponse, DnsRecordUpdate, PaginatedResponse
from app.services import DnsRecordService

router = APIRouter(tags=["records"])


@router.get("/hosted-zones/{zone_id}/records", response_model=PaginatedResponse[DnsRecordResponse])
def list_records(
    zone_id: int,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: Optional[str] = Query(None),
    type: Optional[str] = Query(None, alias="type"),
    sort_by: str = Query("name"),
    sort_order: str = Query("asc"),
    _: str = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    service = DnsRecordService(db)
    result = service.list_records(zone_id, page, page_size, search, type, sort_by, sort_order)
    if result is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Hosted zone not found")
    return result


@router.post(
    "/hosted-zones/{zone_id}/records",
    response_model=DnsRecordResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_record(
    zone_id: int,
    body: DnsRecordCreate,
    _: str = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    service = DnsRecordService(db)
    try:
        record = service.create_record(zone_id, body)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Hosted zone not found")
    return record


@router.put("/records/{record_id}", response_model=DnsRecordResponse)
def update_record(
    record_id: int,
    body: DnsRecordUpdate,
    _: str = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    service = DnsRecordService(db)
    try:
        record = service.update_record(record_id, body)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Record not found")
    return record


@router.delete("/records/{record_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_record(
    record_id: int,
    _: str = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    service = DnsRecordService(db)
    if not service.delete_record(record_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Record not found")
    return None
